MedallionShell.Tests/AttachingTests.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Medallion.Shell.Tests
+{
+    [TestClass]
+    public class AttachingTests
+    {
+        [TestMethod]
+        public void TestAttachingToExistingProcess()
+        {
+            var processCommand = Command.Run("SampleCommand", new[] { "sleep", "10000" });
+            var processId = processCommand.ProcessId;
+            Command.TryAttachToProcess(processId, out _)
+                .ShouldEqual(true, "Attaching to process failed.");
+            processCommand.Kill();
+        }
+
+        [TestMethod]
+        public void TestWaitingForAttachedProcessExit()
+        {
+            var processCommand = Command.Run("SampleCommand", new[] { "sleep", "300" });
+            Command.TryAttachToProcess(processCommand.ProcessId, out var attachedCommand)
+                .ShouldEqual(true, "Attaching to process failed.");
+            var commandResult = attachedCommand.Task;
+            commandResult.IsCompleted.ShouldEqual(false, "Task has finished too early.");
+            Thread.Sleep(1000);
+            commandResult.IsCompleted.ShouldEqual(true, "Task has not finished on time.");
+        }
+
+        [TestMethod]
+        public void TestGettingExitCodeFromAttachedProcess()
+        {
+            var processCommand = Command.Run("SampleCommand", new[] { "exit", "16" });
+            Command.TryAttachToProcess(processCommand.ProcessId, out var attachedCommand)
+                .ShouldEqual(true, "Attaching to process failed.");
+            var task = attachedCommand.Task;
+            task.Wait(1000).ShouldEqual(true, "Task has not finished on time.");
+            task.Result.ExitCode.ShouldEqual(16, "Exit code was not correct.");
+        }
+
+        [TestMethod]
+        public void TestAttachingToNonExistingProcess()
+        {
+            var processCommand = Command.Run("SampleCommand", new[] { "exit", "0" });
+            var processId = processCommand.ProcessId;
+            processCommand.Task.Wait(1000).ShouldEqual(true, "Process has not exited, test is inconclusive.");
+            Command.TryAttachToProcess(processId, out _)
+                .ShouldEqual(false, "Attaching succeeded although process has already exited.");
+        }
```

Comment:
Let's add (1) a test case for calling Kill() after attaching to an existing process and (2) a test case for attaching to an existing process using an already-canceled cancellation token. (2) is particularly interesting because the behavior probably needs to be that it kills the process, but a naive implementation could easily make it so that it bailed out too early to do that.

Reply:
👍